File: 2022/24/rust/src/blizzard.rs
// --- Constants ---
const NORTH: char = '^';
const SOUTH: char = 'v';
const WEST:  char = '<';
const EAST:  char = '>';
const NONE:  char = '.';

// --- Blizzards going east and west ---
struct Row {
    east: Vec<bool>,
    west: Vec<bool>
}

impl Row {
    fn new() -> Self {
        let east = Vec::new();
        let west = Vec::new();
        Self { east, west }
    }

    fn add(&mut self, dir: char) {
        match dir {
            EAST => { self.east.push(true);  self.west.push(false); },
            WEST => { self.east.push(false); self.west.push(true);  },
            NORTH | SOUTH | NONE => { self.east.push(false); self.west.push(false); },
            _ => unreachable!()
        }
    }

    fn is_empty(&self, x: usize, time: usize) -> bool {
        let len = self.east.len();
        let wi = (x - 1 + time) % len;
        let ei = (x - 1 + len - (time % len)) % len;
        !self.east[ei] && !self.west[wi]
    }
}

// --- Blizzards going north and south ---
struct Col {
    north: Vec<bool>,
    south: Vec<bool>
}

impl Col {
    fn new() -> Self {
        let north = Vec::new();
        let south = Vec::new();
        Self { north, south }
    }

    fn add(&mut self, dir: char) {
        match dir {
            NORTH => { self.north.push(true);  self.south.push(false); },
            SOUTH => { self.north.push(false); self.south.push(true);  },
            WEST | EAST | NONE  => { self.north.push(false); self.south.push(false); },
            _ => unreachable!()
        }
    }

    fn is_empty(&self, y: usize, time: usize) -> bool {
        let len = self.north.len();
        let ni = (y - 1 + time) % len;
        let si = (y - 1 + len - (time % len)) % len;
        !self.north[ni] && !self.south[si]
    }
}

// --- Blizzard struct ---
pub struct Blizzard {
    cols: Vec<Col>,
    rows: Vec<Row>
}

impl Blizzard {
    pub fn new() -> Self {
        let cols = Vec::new();
        let rows = Vec::new();
        Self { cols, rows }
    }

    pub fn add(&mut self, line: &str) {
        let mut row = Row::new();
        while self.cols.len() < line.len() { self.cols.push(Col::new()); }
        for (x, d) in line.chars().enumerate() {
            self.cols.get_mut(x).unwrap().add(d);
            row.add(d);
        }
        self.rows.push(row);
    }

    pub fn is_empty(&self, x: usize, y: usize, time: usize) -> bool {
        self.cols[x - 1].is_empty(y, time) && self.rows[y - 1].is_empty(x, time)
    }
}

File: 2022/24/javascript/blizzard.mjs
// --- Constants ---
const NORTH = '^';
const SOUTH = 'v';
const WEST  = '<';
const EAST  = '>';

// --- Blizzards going east and west ---
function Row() {
	this.east = [];
	this.west = [];

	this.add = function(dir) {
		switch(dir) {
			case EAST: this.east.push(true);  this.west.push(false); break;
			case WEST: this.east.push(false); this.west.push(true);  break;
			default:   this.east.push(false); this.west.push(false);
		}
	};

	this.isEmpty = function(x, time) {
		const len = this.east.length;
		const wi = (x - 1 + time) % len;
		const ei = (x - 1 + len - (time % len)) % len;
		return !this.east[ei] && !this.west[wi];
	};
}

// --- Blizzards going north and south ---
function Col() {
	this.north = [];
	this.south = [];

	this.add = function(dir) {
		switch(dir) {
			case NORTH: this.north.push(true);  this.south.push(false); break;
			case SOUTH: this.north.push(false); this.south.push(true);  break;
			default:    this.north.push(false); this.south.push(false);
		}
	};

	this.isEmpty = function(y, time) {
		const len = this.north.length;
		const ni = (y - 1 + time) % len;
		const si = (y - 1 + len - (time % len)) % len;
		return !this.north[ni] && !this.south[si];
	};
}

// --- Blizzard object ---
function Blizzard() {
	this.cols = [];
	this.rows = [];

	this.add = function(line) {
		const chars = line.split("");
		let row = new Row();
		while (this.cols.length < chars.length) this.cols.push(new Col());
		for (let x = 0; x < chars.length; x++) {
			const d = chars[x];
			let col = this.cols[x];
			col.add(d);
			row.add(d);
		}
		this.rows.push(row);
	};

	this.isEmpty = function(x, y, time) {
		return this.cols[x - 1].isEmpty(y, time) && this.rows[y - 1].isEmpty(x, time);
	};
}

export default Blizzard;
